content/posts/2024-11-05-flang-new.md

```diff
@@ -0,0 +1,538 @@
+MLIR's flexibility means that there is no need for HLFIR to have a `hlfir.array`
+which duplicates `fir.array`, unless HLFIR wanted to add extra information.
+
+The next step is to convert the HLFIR parts into FIR, making use of MLIR's
+`arith` (arithmetic) dialect as well.
+
+```mlir
+module attributes {<...>} {
+  func.func @_QPfoo(<...>) {
+    <...>
+    %c3_i32 = arith.constant 3 : i32
+    %7 = fir.convert %0 : (!fir.ref<!fir.box<!fir.array<?xf32>>>) -> !fir.ref<!fir.box<none>>
+    %8 = fir.convert %5 : (!fir.box<!fir.array<?xf32>>) -> !fir.box<none>
+    %9 = fir.convert %6 : (!fir.ref<!fir.char<1,17>>) -> !fir.ref<i8>
+    %10 = fir.call @_FortranAAssign(%7, %8, %9, %c3_i32) : (!fir.ref<!fir.box<none>>, !fir.box<none>, !fir.ref<i8>, i32) -> none
+    return
+  }
+<...>
+}
+```
+Then this bundle of MLIR dialects is converted into purely LLVM IR:
+```mlir
+define void @foo_(ptr %0, ptr %1) {
+  <...>
+  store { ptr, i64, i32, i8, i8, i8, i8, [1 x [3 x i64]] } %37, ptr %3, align 8
+  call void @llvm.memcpy.p0.p0.i32(ptr %5, ptr %4, i32 48, i1 false)
+  %38 = call {} @_FortranAAssign(ptr %5, ptr %3, ptr @_QQclX2F6170702F6578616D706C652E66393000, i32 3)
+  ret void
+}
+<...>
+```
+This LLVM IR passes through the standard optimisation pipeline that LLVM IR produced
+by Clang would. Eventually being converted into target specific Machine IR
+(MIR) and finally machine code.
+
+At each of the stages, the optimisations most suited to that stage are done.
+For example, while we are in HLFIR we could optimise array accesses because in
+HLFIR we have information about how the Fortran language treats those.
+
+If we were to do this later in LLVM IR it would be much more difficult. Either
+the information would be lost or incomplete, or we would be at a stage in the pipeline
+where we cannot assume that we started with a specific source language.
+
+# Flang Takes Driving Lessons
+
+**Note:** This section paraphrases material from
+["Flang Drivers"](https://github.com/llvm/llvm-project/blob/main/flang/docs/FlangDriver.md).
+If you want more detail please refer to that document.
+
+"Driver" in a compiler context means the part of the compiler that decides
+how to handle a set of options. For instance, when you use the option `-march=armv8a+memtag`,
+something in Flang knows that you want to compile for Armv8.0-a with the memory
+tagging extension enabled.
+
+`-march=` is an example of a "compiler driver" option. These options are what users
+give to the compiler. There is actually a second driver after this, confusingly
+called the "frontend" driver, despite being behind the scenes.
+
+In this case the "compiler driver" is `flang` and the "frontend driver" is
+`flang -fc1` (they are in the same binary).
+
+The main reason they are split into 2 is so that the "compiler driver" can have
+option names that do not change over time and the "frontend driver" can change
+its option names as suits the developers of the compiler.
```

Comment:
Another argument (well, same argument, different phrasing):
* compiler driver is for end-users who want to generate executables,
* frontend driver is for compiler developers interested in the internals of the Frontend/Flang itself.

Reply:
I've gone with a mix of both.